tests/zkevm/test_worst_stateful_opcodes.py

```diff
@@ -475,3 +479,208 @@ def test_worst_extcodecopy_warm(
+    ) + Op.RETURN(0, Op.EXTCODESIZE(selfdestructable_contract_addr))
+    initcode_address = pre.deploy_contract(code=initcode)
+
+    # Create a factory that deployes a new SELFDESTRUCT contract instance pre-funded depending on
+    # the value_bearing parameter. We use CREATE2 so the caller contract can easily reproduce
+    # the addresses in a loop for CALLs.
+    factory_code = (
+        Op.EXTCODECOPY(
+            address=initcode_address,
+            dest_offset=0,
+            offset=0,
+            size=Op.EXTCODESIZE(initcode_address),
+        )
+        + Op.MSTORE(
+            0,
+            Op.CREATE2(
+                value=1 if value_bearing else 0,
+                offset=0,
+                size=Op.EXTCODESIZE(initcode_address),
+                salt=Op.SLOAD(0),
+            ),
+        )
+        + Op.SSTORE(0, Op.ADD(Op.SLOAD(0), 1))
+        + Op.RETURN(0, 32)
+    )
+    factory_address = pre.deploy_contract(code=factory_code, balance=10**18)
+
+    factory_caller_code = Op.CALLDATALOAD(0) + While(
+        body=Op.POP(Op.CALL(address=factory_address)),
+        condition=Op.PUSH1(1) + Op.SWAP1 + Op.SUB + Op.DUP1 + Op.ISZERO + Op.ISZERO,
+    )
+    factory_caller_address = pre.deploy_contract(code=factory_caller_code)
+
+    gas_costs = fork.gas_costs()
+    intrinsic_gas_cost_calc = fork.transaction_intrinsic_cost_calculator()
+    loop_cost = (
+        gas_costs.G_KECCAK_256  # KECCAK static cost
+        + math.ceil(85 / 32) * gas_costs.G_KECCAK_256_WORD  # KECCAK dynamic cost for CREATE2
+        + gas_costs.G_VERY_LOW * 3  # ~MSTOREs+ADDs
+        + gas_costs.G_COLD_ACCOUNT_ACCESS  # CALL to self-destructing contract
+        + gas_costs.G_SELF_DESTRUCT
+        + 30  # ~Gluing opcodes
+    )
+    num_contracts = (
+        # Base available gas = GAS_LIMIT - intrinsic - (out of loop MSTOREs)
+        attack_gas_limit - intrinsic_gas_cost_calc() - gas_costs.G_VERY_LOW * 4
+    ) // loop_cost
+
+    contracts_deployment_tx = Transaction(
+        to=factory_caller_address,
+        gas_limit=env.gas_limit,
+        gas_price=10**9,
+        data=Hash(num_contracts),
+        sender=pre.fund_eoa(),
+    )
+
+    code = (
+        # Setup memory for later CREATE2 address generation loop.
+        # 0xFF+[Address(20bytes)]+[seed(32bytes)]+[initcode keccak(32bytes)]
+        Op.MSTORE(0, factory_address)
+        + Op.MSTORE8(32 - 20 - 1, 0xFF)
+        + Op.MSTORE(32, 0)
+        + Op.MSTORE(64, initcode.keccak256())
+        # Main loop
+        + While(
+            body=Op.POP(Op.CALL(address=Op.SHA3(32 - 20 - 1, 85)))
+            + Op.MSTORE(32, Op.ADD(Op.MLOAD(32), 1)),
+            # Stop before we run out of gas for the whole tx execution.
+            # The value was found by trial-error rounded to the next 1000 multiple.
+            condition=Op.GT(Op.GAS, 12_000),
+        )
+        + Op.SSTORE(0, 42)  # Done for successful tx execution assertion below.
```

Comment:
@jochem-brouwer, here I am experimenting to have successful-tx execution detection via:
- Early stop of the `While` using an amount that is fixed and found by the test creator. The way I did this practically was some binary search.
- I do a `SSTORE(0, 42)` that is later added as a post-state assertion.

The found magic number (i.e. `12_000` in this case), should not depend on the env `gas_limit`, so I think feels OK.

If somebody does a refactor of the test, or there're gas repricings then this test will break, but  feels like fixing it is quite easy assuming my code comment is clear enough.

The overhead is the `condition` execution plus the `SSTORE`.

_Looks_ simple enough and with reasonable tradeoffs, without having to put a very complicated formula that is very brittle.

Reply:
I think the current check is very good and it should be kept! It will enforce that the tx cannot go OOG (could be due to wrong code / stack overflow/underflows) and that thus the "loop" has ran at least once. It is not perfect but this is a great improvement :smile: :+1: 

Thinking about full support for other forks which might change gas schedules (and might thus also break this test) is maybe also a bit out-of-scope for now. But nevertheless good to keep in mind.